text/XXX-unwind-FFI.md

```diff
@@ -0,0 +1,286 @@
+
+This RFC does, however, provide a tool that will enable users
+to provide this guarantee themselves.
+We will stabilize the `!#[panic_runtime]` annotation,
+which [designates a crate as the provider of the final product's panic runtime]
+(less-unwinding).
+Additionally, the standard library's current `panic=unwind` runtime crate,
+`libpanic_unwind`, which is compatible with native C++ style exceptions,
+will be provided under another name (such as `libpanic_native`)
+that guarantees the implementation will remain compatible with C++ exceptions
+(while the implementation of `libpanic_unwind` itself may change
+to no longer maintain that compatibility).
+
+In order for Cargo users to be able
+to specify the C++ compatible `panic_runtime` implementation,
+a new optional value, runtime, will be added to the `profile.foo.panic` option:
+
+```toml
+[profile.dev]
+panic = { strategy = 'unwind', runtime = 'native' }
+```
+
+`panic.runtime` may only be specified if the panic strategy is `unwind`.
+And just as there may only be one strategy
+in the dependency graph for a final product,
+there may only be one `runtime`.
+
+For now, the only valid `runtime` keys will be:
+
+* `default` - identical to `panic = 'unwind'` with no `runtime` selected.
+  This will use Rust's default unwinding runtime,
+  unless another crate in the final product's dependency graph
+specifies `panic = 'abort'`,
+  in which case the `abort` strategy will take precedence as usual.
+  which is not guaranteed to be compatible with native exceptions.
+* `native` - as discussed above, this will preserve the behavior of the current
+  implementation of `libpanic_unwind`.
+* `crate` - indicates that a non-`std` crate will use `#![panic_runtime]` to
+  provide the runtime.
+* `self` - indicates that this crate itself provides `#![panic_runtime]`.
+
+If the `native` or `crate` key is specified
+anywhere in a final product's dependency graph,
+no crate in that dependency graph may specify the `panic = abort` strategy;
+this mismatch will cause the build to fail.
+
+The function annotation `#[unwind(allowed)]` will only be permitted in crates
+that specify a non-default `panic.runtime`.
+
+Crates that do not use the `profile.foo.panic` option at all
+will remain compatible with any `profile.foo.panic` configuration
+used to generate the final product.
+
+For non-Cargo users, equivalent `rustc` flags will be provided
+(which will be how Cargo itself implements the option).
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Unwinding for functions with the `#[unwind(allowed)]` annotation
+is performed as if the function were not marked `extern`.
+This annotation is not permitted on functions not marked `extern`.
+It has no observable effect unless the marked function `panic`s
+(e.g. it has no observable effect
+when a function returns normally or enters an infinite loop).
+The LLVM IR for such functions must not be marked `nounwind`.
```

Comment:
This is an implementation detail and should not be normative. If the compiler can prove that no function in the call graph could possibly unwind, it is perfectly fine for the function to be marked as `nounwind` nevertheless.

Reply:
That's fair enough; I thought I had received feedback on my first draft that I should explicitly include mention of not emitting `nounwind`, but I can't actually find that comment so I may simply be mis-remembering.

```suggestion
```